pkg/registry/common/connect/nse_server.go

```diff
@@ -0,0 +1,120 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package connect
+
+import (
+	"context"
+	"time"
+
+	"github.com/networkservicemesh/sdk/pkg/tools/serialize"
+
+	"github.com/golang/protobuf/ptypes/empty"
+	"github.com/networkservicemesh/api/pkg/api/registry"
+	"google.golang.org/grpc"
+
+	"github.com/networkservicemesh/sdk/pkg/registry/common/clienturl"
+	"github.com/networkservicemesh/sdk/pkg/registry/core/adapters"
+)
+
+type nseCacheEntry struct {
+	expiryTime time.Time
+	client     registry.NetworkServiceEndpointRegistryClient
+}
+
+type connectNSEServer struct {
+	dialOptions       []grpc.DialOption
+	clientFactory     func(ctx context.Context, cc grpc.ClientConnInterface) registry.NetworkServiceEndpointRegistryClient
+	cache             map[string]*nseCacheEntry
+	ctx               context.Context
+	connectExpiration time.Duration
+	executor          serialize.Executor
+}
+
+// NewNetworkServiceEndpointRegistryServer creates new connect NetworkServiceEndpointEndpointRegistryServer with specific chain context, registry client factory and options
+// that allows connecting to other registries via passed clienturl.
+func NewNetworkServiceEndpointRegistryServer(
+	chainContext context.Context,
+	clientFactory func(ctx context.Context, cc grpc.ClientConnInterface) registry.NetworkServiceEndpointRegistryClient,
+	options ...Option) registry.NetworkServiceEndpointRegistryServer {
+	r := &connectNSEServer{
+		clientFactory:     clientFactory,
+		cache:             map[string]*nseCacheEntry{},
+		ctx:               chainContext,
+		connectExpiration: time.Minute,
+	}
+	for _, o := range options {
+		o.apply(r)
+	}
+	return r
+}
+
+func (c *connectNSEServer) Register(ctx context.Context, ns *registry.NetworkServiceEndpoint) (*registry.NetworkServiceEndpoint, error) {
+	return c.connect(ctx).Register(ctx, ns)
+}
+
+func (c *connectNSEServer) Find(q *registry.NetworkServiceEndpointQuery, s registry.NetworkServiceEndpointRegistry_FindServer) error {
+	return adapters.NetworkServiceEndpointClientToServer(c.connect(s.Context())).Find(q, s)
+}
+
+func (c *connectNSEServer) Unregister(ctx context.Context, ns *registry.NetworkServiceEndpoint) (*empty.Empty, error) {
+	return c.connect(ctx).Unregister(ctx, ns)
+}
+
+func (c *connectNSEServer) cleanup() {
+	var deleteList []string
+	for k, v := range c.cache {
+		if time.Until(v.expiryTime) <= 0 {
+			deleteList = append(deleteList, k)
+		}
+	}
+	for _, item := range deleteList {
+		delete(c.cache, item)
+	}
+}
+
+func (c *connectNSEServer) connect(ctx context.Context) registry.NetworkServiceEndpointRegistryClient {
+	defer c.executor.AsyncExec(c.cleanup)
+	key := ""
+	if url := clienturl.ClientURL(ctx); url != nil {
+		key = url.String()
+	}
+	var result registry.NetworkServiceEndpointRegistryClient
+	<-c.executor.AsyncExec(func() {
```

Comment:
OK... we've moved the global synchronicity to an executor... but its still present.  This could be done similarly to how NetworkService connect does it *without* hitting global locks on every call.

Reply:
The main difference between NetworkService connect and registry connect is that registry connect can not just know when the registry can delete a client to free resources. So that's why we need to use some strategy for finding and cleaning outdated resources. So, in this case, we need to use a `lock` because of finding and cleaning should be covered by a critical section to prevent some extrimal race conditions. In this case, executor looks good to me because it has smart goroutine spawning behavior.

Reply:
Don't registry entries have an expireTime?

Does the critical section need to be scoped globally?

Reply:
> Don't registry entries have an expireTime?

Yes NSE has expiration_time but how you plan to use it here? 

For example, the user doing these steps:

```
1. Find nse1@domain1
2. Register nse2@floating.domain
3. Register nse3@floating.domain
4. Unregister nse4@domain1
```

So here we have two clients for _floating.domain_ and for _domain1_. Question: what the strategy you are seeing to use here?

Reply:
> Does the critical section need to be scoped globally?

Not sure that pattern from networkservice.connect is suitable here because we need to do two things in one critical section:
1. Iterate whole map of clients
2. Delete expired clients

Reply:
Or just use time.Timer :)  Reduces the scope dramatically :)

Reply:
> Or just use time.Timer :) Reduces the scope dramatically :)

With timers, we have the same issue:
```
timer tick:
1. Iterate whole map of clients
<---race condition here for sync map
2. Delete expired clients
```
```
Register/Find/Unregister:
1. Lookup in cache
```
Also, timers will take some goroutine when in the case when connect is not used executor can kill goroutine and do not use machine resources to check empty map :)

Reply:
Each Registry Entry we are tracking has its own expirationTimer... and could thus have its own timer.  The *event* being the expiration of the Registry entry.  If you track things similar to how we handle connections in Network Service connect, you should be able to avoid global locks, and global traversals.

Reply:
@edwarnicke ok, how you want to handle the case when user Finds NSE with expiration_time = 1 year? Do you suggest never a a close connection in this case?

Reply:
Finally should be fixed.